src/ILCompiler.ReadyToRun/src/Compiler/ReadyToRunMetadataFieldLayoutAlgorithm.cs

```diff
@@ -4,44 +4,241 @@
         /// <summary>
-        /// CoreCLR DomainLocalModule::OffsetOfDataBlob() / sizeof(void *)
+        /// Map from EcmaModule instances to field layouts within the individual modules.
         /// </summary>
-        private const int DomainLocalModuleDataBlobOffsetAsIntPtrCount = 6;
+        private ModuleFieldLayoutMap _moduleFieldLayoutMap;
 
-        /// <summary>
-        /// CoreCLR ThreadLocalModule::OffsetOfDataBlob() / sizeof(void *)
-        /// </summary>
-        private const int ThreadLocalModuleDataBlobOffsetAsIntPtrCount = 3;
-
-        private LayoutInt _initialNonGcStaticsOffset;
+        public ReadyToRunMetadataFieldLayoutAlgorithm()
+        {
+            _moduleFieldLayoutMap = new ModuleFieldLayoutMap();
+        }
 
-        public ReadyToRunMetadataFieldLayoutAlgorithm(TargetDetails target, int numberOfTypesInModule)
+        public override ComputedStaticFieldLayout ComputeStaticFieldLayout(DefType defType, StaticLayoutKind layoutKind)
         {
-            _initialNonGcStaticsOffset = new LayoutInt(DomainLocalModuleDataBlobOffsetAsIntPtrCount * target.PointerSize + numberOfTypesInModule);
+            ComputedStaticFieldLayout layout = new ComputedStaticFieldLayout();
+            if (defType is EcmaType ecmaType)
+            {
+                // ECMA types are the only ones that can have statics
+                ModuleFieldLayout moduleFieldLayout = _moduleFieldLayoutMap.GetOrCreateValue(ecmaType.EcmaModule);
+                layout.GcStatics = moduleFieldLayout.GcStatics;
+                layout.NonGcStatics = moduleFieldLayout.NonGcStatics;
+                layout.ThreadGcStatics = moduleFieldLayout.ThreadGcStatics;
+                layout.ThreadNonGcStatics = moduleFieldLayout.ThreadNonGcStatics;
+                moduleFieldLayout.TypeToFieldMap.TryGetValue(defType, out layout.Offsets);
+            }
+            return layout;
         }
 
-        protected override void PrepareRuntimeSpecificStaticFieldLayout(TypeSystemContext context, ref ComputedStaticFieldLayout layout)
+        /// <summary>
+        /// Map from modules to their static field layouts.
+        /// </summary>
+        private class ModuleFieldLayoutMap : LockFreeReaderHashtable<EcmaModule, ModuleFieldLayout>
         {
-            layout.NonGcStatics.Size = _initialNonGcStaticsOffset;
-            layout.GcStatics.Size = LayoutInt.Zero;
-            layout.ThreadNonGcStatics.Size = LayoutInt.Zero;
-            layout.ThreadGcStatics.Size = LayoutInt.Zero;
+            /// <summary>
+            /// CoreCLR DomainLocalModule::OffsetOfDataBlob() / sizeof(void *)
+            /// </summary>
+            private const int DomainLocalModuleDataBlobOffsetAsIntPtrCount = 6;
+
+            /// <summary>
+            /// CoreCLR ThreadLocalModule::OffsetOfDataBlob() / sizeof(void *)
+            /// </summary>
+            private const int ThreadLocalModuleDataBlobOffsetAsIntPtrCount = 3;
+
+            public ModuleFieldLayoutMap()
```

Comment:
C# will generate an empty default constructor for us. This can be deleted.

Reply:
Deleted.